onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu

```diff
@@ -0,0 +1,775 @@
+          static_cast<scalar_t>(smem[threadIdx.x] + prevHighProb);
+      scalar_t prevBucket = static_cast<scalar_t>(
+          threadIdx.x == 0 ? prevHighProb
+                          : smem[threadIdx.x - 1] + prevHighProb);
+      bool inBucket =
+          (cat < categories) &&
+          (!(sample >= curBucket) &&
+          (sample >= prevBucket) &&
+          (dist_val > zero));
+      if (inBucket) {
+        // We're done; we have the sample
+        // Torch indices are 1-based
+        atomicMax(&foundPos, cat);
+        found = true;
+      }
+      // Store the previous scan's high value for future use
+      prevHighProb = prevHighProb + smem[blockDim.x - 1];
+      __syncthreads();
+    }
+    if (threadIdx.x == 0) {
+      if (found) {
+          dest[curDist] = foundPos;
+      } else {
+        // This should address a rare bug where we don't select a valid index. This likely occurs when
+        // due to floating point arithmetic rounding errors, our cumulative sum does not add up to 1, but
+        // and our uniform sample is greater than this value. In this case we likely have unitialized memory
+        // in dest[curDist]. So basically we will loop through the distribution and pick the largest index
+        // where the distribution is non-zero. This is obviously terribly inefficient, but due to the
+        // rarity in which this occurs, this should not be an issue.
+        for (int cat = categories - 1; cat >= 0; --cat) {
+          if (dist[curDist * stride_dist + cat * stride_categories] > zero) {
+            dest[curDist] = cat;
+            break;
+          }
+        }
+      }
+    }
+  }
+
+  // update presence mask
+  int index = blockIdx.x * blockDim.x + threadIdx.x;
+  if (index >= distributions * categories) {
+    return;
+  }
+  int dist_idx = index / categories;
+  int cat_idx = index % categories;
+  if (dest[dist_idx] == cat_idx) {
+    d_presence_mask[index] = 1;
+  }
+}
+
+// Only support n_sample = 1
+void TorchMultinomialKernelLauncher(float* d_input,
+                                    float* d_sampled,
+                                    int64_t* d_output,
+                                    int batch_size,
+                                    int vocab_size,
+                                    int* d_presence_mask,
+                                    cudaStream_t stream)
+{
+  // Store the props in class variables
+  int device;
+  cudaGetDevice(&device);
+  cudaDeviceProp props;
+  cudaGetDeviceProperties(&props, device);
+
+  int numSM = props.multiProcessorCount;
+  int maxThreads = props.maxThreadsPerBlock;
+  int warp_size = 32; //at::cuda::warp_size();
+  int requiredWarps = (vocab_size + warp_size - 1) / warp_size;
+  int requiredThreads = std::min(maxThreads, requiredWarps * warp_size);
+  int requiredShared = requiredThreads * sizeof(float);
+
+  // bugbug: randomize d_sampled
+  dim3 block(requiredThreads);
+  dim3 grid(std::min(batch_size, numSM * 4));
```

Comment:
>batch_size, numSM * 4 [](http://example.com/codeflow?start=21&length=21)

why not juse use batch_size?

Reply:
this function is basically copied from pytorch cuda implementation so I guess there might be some advantages here